block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -53,14 +53,71 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
       (blockHash.size == BlockHash.Length && dagSet.contains(blockHash)).pure[F]
 
     def children(blockHash: BlockHash): F[Option[Set[BlockHash]]] =
-      childMap.get(blockHash).pure[F]
+      childMap.get(blockHash).map(_ intersect dagSet).pure[F]
 
     def latestMessageHash(validator: Validator): F[Option[BlockHash]] =
       latestMessagesMap.get(validator).pure[F]
 
     def latestMessageHashes: F[Map[Validator, BlockHash]] = latestMessagesMap.pure[F]
 
-    def invalidBlocks: F[Set[BlockMetadata]] = invalidBlocksSet.pure[F]
+    def invalidBlocks: F[Set[BlockMetadata]] =
+      invalidBlocksSet.filter(m => dagSet.contains(m.blockHash)).pure[F]
+
+    override def truncate(
+        latestMessages: Map[Validator, BlockHash]
+    ): F[BlockDagRepresentation[F]] = {
+      final case class HeightView(
+          height: Long,
+          excess: Iterable[ByteString],
+          inView: Iterable[ByteString]
+      )
+      for {
+        lmMetas <- latestMessages.values.toStream
+                    .traverse(lookup)
+                    .map(_.map(_.get).toSet)
```

Comment:
You can use _unsafe_ lookup here. Please avoid using _.get_ on an Option type directly because becomes the same as using _null_.